File: platform/util/src/com/intellij/jna/JnaLoader.java
// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.jna;

import com.intellij.openapi.diagnostic.Logger;
import com.intellij.util.system.OS;
import com.sun.jna.Native;
import org.jetbrains.annotations.NotNull;

public final class JnaLoader {
  private static Boolean ourJnaLoaded = null;

  public static synchronized void load(@NotNull Logger logger) {
    if (ourJnaLoaded == null) {
      ourJnaLoaded = Boolean.FALSE;

      OS os = OS.CURRENT;
      if (os == OS.Windows && Boolean.getBoolean("ide.native.launcher")) {
        // temporary fix for JNA + `SetDefaultDllDirectories` DLL loading issue (IJPL-157390)
        String winDir = System.getenv("SystemRoot");
        if (winDir != null) {
          String path = System.getProperty("jna.platform.library.path");
          path = (path == null ? "" : path + ';') + winDir + "\\System32";
          System.setProperty("jna.platform.library.path", path);
        }
      }

      try {
        long t = System.currentTimeMillis();
        int ptrSize = Native.POINTER_SIZE;
        t = System.currentTimeMillis() - t;
        logger.info("JNA library (" + (ptrSize << 3) + "-bit) loaded in " + t + " ms");
        ourJnaLoaded = Boolean.TRUE;
      }
      catch (Throwable t) {
        logger.warn(
          "Unable to load JNA library (" + os + '/' + os.version() + ", jna.boot.library.path=" + System.getProperty("jna.boot.library.path") + ')',
          t);
      }
    }
  }

  public static synchronized boolean isLoaded() {
    if (ourJnaLoaded == null) {
      load(Logger.getInstance(JnaLoader.class));
    }
    return ourJnaLoaded;
  }
}

File: platform/jewel/int-ui/int-ui-standalone/src/main/kotlin/org/jetbrains/jewel/intui/standalone/window/JnaLoader.kt
// Copyright 2000-2026 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.jewel.intui.standalone.window

import com.sun.jna.Native
import java.util.logging.Level
import java.util.logging.Logger
import kotlin.system.measureTimeMillis
import org.jetbrains.annotations.ApiStatus
import org.jetbrains.jewel.foundation.InternalJewelApi

/**
 * Lazily initialises the JNA native library and exposes whether the load succeeded.
 *
 * Calls are thread-safe: both [load] and [isLoaded] are `@Synchronized`.
 */
@ApiStatus.Internal
@InternalJewelApi
public object JnaLoader {
    private var loaded: Boolean? = null
    private val logger = Logger.getLogger(JnaLoader::class.java.simpleName)

    /** Loads the JNA native library if not already loaded. On failure, logs a warning and leaves [isLoaded] false. */
    @Synchronized
    public fun load() {
        if (loaded == null) {
            loaded = false
            try {
                val time = measureTimeMillis { Native.POINTER_SIZE }
                logger.info("JNA library (${Native.POINTER_SIZE shl 3}-bit) loaded in $time ms")
                loaded = true
            } catch (@Suppress("TooGenericExceptionCaught") t: Throwable) {
                logger.log(
                    Level.WARNING,
                    "Unable to load JNA library(os=${
                        System.getProperty("os.name")
                    } ${System.getProperty("os.version")}, jna.boot.library.path=${
                        System.getProperty("jna.boot.library.path")
                    })",
                    t,
                )
            }
        }
    }

    /** Whether the JNA library was successfully loaded. Triggers [load] on first access. */
    @get:Synchronized
    public val isLoaded: Boolean
        get() {
            if (loaded == null) {
                load()
            }
            return loaded ?: false
        }
}
